core/src/main/java/com/linecorp/armeria/internal/common/DefaultSplitHttpResponse.java

```diff
@@ -0,0 +1,358 @@
+        }
+
+        private void request0(long n) {
+            final Subscription upstream = this.upstream;
+            if (upstream == null) {
+                pendingRequests = LongMath.saturatedAdd(n, pendingRequests);
+            } else {
+                upstream.request(n);
+            }
+        }
+
+        @Override
+        public void cancel() {
+            if (cancelCalled) {
+                return;
+            }
+            cancelCalled = true;
+            if (!notifyCancellation) {
+                downstream = NoopSubscriber.get();
+            }
+            maybeCompleteHeaders(null);
+            final Subscription upstream = this.upstream;
+            if (upstream != null) {
+                upstream.cancel();
+            }
+        }
+
+        @Override
+        public void onNext(HttpObject httpObject) {
+            if (httpObject instanceof ResponseHeaders) {
+                final ResponseHeaders headers = (ResponseHeaders) httpObject;
+                final HttpStatus status = headers.status();
+                if (status.isInformational()) {
+                    // Ignore informational headers
+                    upstream.request(1);
+                } else {
+                    headersFuture.doComplete(headers);
+                }
+                return;
+            }
+
+            if (httpObject instanceof HttpHeaders) {
+                final HttpHeaders trailers = (HttpHeaders) httpObject;
+                completeTrailers(trailers);
+                return;
+            }
+
+            final Subscriber<? super HttpData> downstream = this.downstream;
+            assert downstream != null;
+            assert httpObject instanceof HttpData;
+            final HttpData data = (HttpData) httpObject;
+            wroteAny = true;
+            downstream.onNext(data);
+        }
+
+        /**
+         * Completes the specified trailers.
+         */
+        private void completeTrailers(HttpHeaders trailers) {
+            HeadersFuture<HttpHeaders> trailersFuture = DefaultSplitHttpResponse.this.trailersFuture;
+            if (trailersFuture != null) {
+                trailersFuture.doComplete(trailers);
+                return;
+            }
+
+            trailersFuture = new HeadersFuture<>();
+            if (trailersFutureUpdater.compareAndSet(DefaultSplitHttpResponse.this, null, trailersFuture)) {
+                trailersFuture.doComplete(trailers);
+            } else {
+                DefaultSplitHttpResponse.this.trailersFuture.doComplete(trailers);
+            }
+        }
+
+        @Override
+        public void onError(Throwable cause) {
+            maybeCompleteHeaders(cause);
+            final Subscriber<? super HttpData> downstream = this.downstream;
+            if (downstream == null) {
+                this.cause = cause;
+            } else {
+                downstream.onError(cause);
+                if (cause instanceof CancelledSubscriptionException) {
```

Comment:
Isn't this `if (notifyCancellation)` ?

Reply:
Yeah, I think that is the same effect. Otherwise, I think we can remove the `if` condition.
There is no harm to set `NoopSubscriber` to `downstream` after `onError()` is invoked.